File: 2300-construct-string-with-repeat-limit/construct-string-with-repeat-limit.c
char* repeatLimitedString(char* s, int repeatLimit) {
    // Count character frequencies
    int count[26] = {0};
    int len = strlen(s);
    
    // Count frequencies
    for (int i = 0; i < len; i++) {
        count[s[i] - 'a']++;
    }
    
    // Allocate result string (maximum possible length)
    char* result = (char*)malloc((len + 1) * sizeof(char));
    int resultIndex = 0;
    result[0] = '\0';
    
    // Process from largest to smallest character
    for (int i = 25; i >= 0; ) {
        // Skip if current character is exhausted
        if (count[i] == 0) {
            i--;
            continue;
        }
        
        // Count consecutive characters at the end of result
        int consecutiveCount = 0;
        if (resultIndex > 0) {
            for (int j = resultIndex - 1; j >= 0 && result[j] == (char)(i + 'a'); j--) {
                consecutiveCount++;
            }
        }
        
        // Determine how many characters we can add
        int toAdd = (consecutiveCount < repeatLimit) ? 
            (consecutiveCount == 0 ? 
                (count[i] < repeatLimit ? count[i] : repeatLimit) : 
                (repeatLimit - consecutiveCount)) : 0;
        
        // Add characters
        for (int j = 0; j < toAdd; j++) {
            result[resultIndex++] = (char)(i + 'a');
            count[i]--;
        }
        result[resultIndex] = '\0';
        
        // If all of current character used, move to next
        if (count[i] == 0) {
            i--;
            continue;
        }
        
        // Find next smaller character to break sequence
        int j = i - 1;
        while (j >= 0 && count[j] == 0) {
            j--;
        }
        
        // If no smaller character, we're done
        if (j < 0) {
            break;
        }
        
        // Add one character from next smaller character
        result[resultIndex++] = (char)(j + 'a');
        result[resultIndex] = '\0';
        count[j]--;
        
        // If no more of this character, move to next
        if (count[j] == 0) {
            j--;
        }
        
        // Reset where we're looking
        i = (i > j) ? i : j;
    }
    
    return result;
}

File: 2300-construct-string-with-repeat-limit/construct-string-with-repeat-limit.java
class Solution {
    public String repeatLimitedString(String s, int repeatLimit) {
        // Count character frequencies
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }
        
        // StringBuilder for result
        StringBuilder result = new StringBuilder(s.length());
        
        // Start from the largest character
        for (int i = 25; i >= 0; ) {
            // If current character is exhausted, move to next
            if (count[i] == 0) {
                i--;
                continue;
            }
            
            // Determine how many of current character we can add
            int toAdd = Math.min(count[i], 
                result.length() > 0 && (char)(i + 'a') == result.charAt(result.length() - 1) 
                ? repeatLimit - getConsecutiveCount(result) 
                : repeatLimit
            );
            
            // Add current character
            for (int j = 0; j < toAdd; j++) {
                result.append((char)(i + 'a'));
            }
            count[i] -= toAdd;
            
            // If we've added all of current character
            if (count[i] == 0) {
                i--;
                continue;
            }
            
            // Try to break the sequence with a smaller character
            int j = i - 1;
            while (j >= 0 && count[j] == 0) {
                j--;
            }
            
            // If no smaller character available, we're done
            if (j < 0) {
                break;
            }
            
            // Add one smaller character
            result.append((char)(j + 'a'));
            count[j]--;
            
            // If no more of this character, move to next
            if (count[j] == 0) {
                j--;
            }
            
            // Reset where we're looking
            i = Math.max(i, j);
        }
        
        return result.toString();
    }
    
    // Helper method to count consecutive characters at the end
    private int getConsecutiveCount(StringBuilder sb) {
        if (sb.length() == 0) return 0;
        
        char last = sb.charAt(sb.length() - 1);
        int count = 1;
        
        for (int i = sb.length() - 2; i >= 0; i--) {
            if (sb.charAt(i) == last) {
                count++;
            } else {
                break;
            }
        }
        
        return count;
    }
}